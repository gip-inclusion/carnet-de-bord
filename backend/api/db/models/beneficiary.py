from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field

from api.db.models.notebook import Notebook


class Beneficiary(BaseModel):
    id: UUID
    email: str | None
    lastname: str
    firstname: str
    caf_number: str | None
    pe_number: str | None
    postal_code: str | None
    city: str | None
    address1: str | None
    address2: str | None
    mobile_number: str | None
    date_of_birth: date
    place_of_birth: str | None
    deployment_id: UUID
    created_at: datetime
    updated_at: datetime
    internal_id: str | None
    notebook: Notebook | None
    # BRSA users may not have an account
    # (account is created on the  first login attempt)
    account_id: UUID | None


class BeneficiaryImport(BaseModel):
    si_id: str = Field(..., title="Identifiant du SI interne")
    firstname: str = Field(..., title="Prénom")
    lastname: str = Field(..., title="Nom")
    date_of_birth: date = Field(..., title="Date de naissance")
    place_of_birth: str | None = Field(None, title="Lieu de naissance")
    phone_number: str | None = Field(None, title="Numéro de téléphone")
    email: str | None = Field(None, title="Email")
    address1: str | None = Field(None, title="Adresse")
    address2: str | None = Field(None, title="Complément d'adresse")
    postal_code: str | None = Field(None, title="Code postal")
    city: str | None = Field(None, title="Ville")
    work_situation: str | None = Field(None, title="Situation professionnelle")
    caf_number: str | None = Field(None, title="Numéro CAF/MSA")
    pe_number: str | None = Field(None, title="Identifiant Pole Emploi")
    right_rsa: str | None = Field(None, title="Droits RSA")
    right_are: bool | None = Field(None, title="Droits ARE")
    right_ass: bool | None = Field(None, title="Droits ASS")
    right_bonus: bool | None = Field(None, title="Droits Bonus")
    right_rqth: bool | None = Field(None, title="Droits RQTH")
    geographical_area: str | None = Field(None, title="Zone de mobilité")
    rome_code_description: str | None = Field(None, title="Emploi recherché")
    education_level: str | None = Field(None, title="Niveau de formation")
    structure_name: str | None = Field(None, title="Structure d'accompagnement")
    advisor_email: str | None = Field(None, title="Accompagnateur référent")
