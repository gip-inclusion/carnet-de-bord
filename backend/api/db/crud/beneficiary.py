from datetime import date
from typing import List
from uuid import UUID

from asyncpg import Record
from asyncpg.connection import Connection
from asyncpg.exceptions import UniqueViolationError

from api.db.crud.notebook import (
    NOTEBOOK_BASE_FIELDS,
    NOTEBOOK_BASE_JOINS,
    parse_notebook_from_record,
)
from api.db.models.beneficiary import Beneficiary, BeneficiaryImport

BENEFICIARY_BASE_QUERY = (
    """SELECT b.*, acc.id as account_id,"""
    + NOTEBOOK_BASE_FIELDS
    + """
    FROM public.beneficiary b
    LEFT JOIN notebook n ON n.beneficiary_id = b.id
    """
    + NOTEBOOK_BASE_JOINS
    + """
    LEFT JOIN public.account acc
    ON acc.beneficiary_id = b.id
"""
)


# see https://betakuang.medium.com/why-postgresqls-on-conflict-cannot-find-my-partial-unique-index-552327b85e1
firstname_lastname_date_of_birth_unique_idx = (
    "(LOWER(trim(firstname)), LOWER(trim(lastname)), date_of_birth, deployment_id)"
)


async def select_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    return await connection.fetch(
        f"""
SELECT (firstname, lastname, date_of_birth, internal_id, deployment_id, id)
FROM beneficiary
WHERE (firstname = $1 AND lastname = $2 AND date_of_birth = $3)
OR (internal_id = $4 AND deployment_id = $5)
        """,
        beneficiary.firstname,
        beneficiary.lastname,
        beneficiary.date_of_birth,
        beneficiary.si_id,
        deployment_id,
    )


async def update_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    await connection.fetchrow(
        f"""
UPDATE beneficiary SET mobile_number = $1
where internal_id = $2
returning id
        """,
        beneficiary.phone_number,
        beneficiary.si_id,
    )


async def insert_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    await connection.fetchrow(
        f"""
INSERT INTO BENEFICIARY (firstname, lastname, internal_id, date_of_birth, deployment_id, mobile_number)
values ($1, $2, $3, $4, $5, $6)
returning id
        """,
        beneficiary.firstname,
        beneficiary.lastname,
        beneficiary.si_id,
        beneficiary.date_of_birth,
        deployment_id,
        beneficiary.phone_number,
    )


async def import_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    result = select_beneficiary(connection, beneficiary, deployment_id)
    existing_rows = await result
    if len(existing_rows) == 0:
        print("inserting new user")
        await insert_beneficiary(connection, beneficiary, deployment_id)
    else:
        print(existing_rows[0][0][4])
        if (
            len(existing_rows) == 1
            and existing_rows[0][0][0] == beneficiary.firstname
            and existing_rows[0][0][1] == beneficiary.lastname
            and existing_rows[0][0][2] == beneficiary.date_of_birth
            and existing_rows[0][0][3] == beneficiary.si_id
        ):
            # Doesn't work somehow
            #     and existing_rows[0][0][4] == deployment_id
            print("updating existing user")
            await update_beneficiary(connection, beneficiary, deployment_id)
        else:
            print("data problem")


async def get_beneficiary_with_query(
    connection: Connection, query: str, *args
) -> Beneficiary | None:

    async with connection.transaction():

        beneficiary = None

        beneficiary_records: List[Record] = await connection.fetch(
            BENEFICIARY_BASE_QUERY + query,
            *args,
        )

        for beneficiary_record in beneficiary_records:

            if beneficiary is None:
                beneficiary = Beneficiary.parse_obj(beneficiary_record)

            if beneficiary_record["n_id"] is not None:
                beneficiary.notebook = await parse_notebook_from_record(
                    beneficiary_record,
                    notebook=beneficiary.notebook,
                )

        return beneficiary


async def get_beneficiary_from_personal_information(
    connection: Connection, firstname: str, lastname: str, birth_date: date
) -> Beneficiary | None:

    return await get_beneficiary_with_query(
        connection,
        # We use LOWER(BTRIM(field)) here to match the index so postgres
        # can use the index rather than scanning the whole table
        "WHERE LOWER(BTRIM(b.firstname)) = LOWER($1) AND "
        "LOWER(BTRIM(b.lastname)) = LOWER($2) AND "
        "b.date_of_birth = $3",
        firstname,
        lastname,
        birth_date,
    )


async def get_beneficiary_by_id(
    connection: Connection, beneficiary_id: UUID
) -> Beneficiary | None:

    return await get_beneficiary_with_query(
        connection, "WHERE b.id = $1", beneficiary_id
    )
