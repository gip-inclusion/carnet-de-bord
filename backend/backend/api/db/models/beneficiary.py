from datetime import date, datetime
from uuid import UUID

from pandas.core.series import Series
from pydantic import BaseModel, EmailStr, Field, ValidationError, validator

from backend.api.db.models.csv import CsvFieldError
from backend.api.db.models.nir import nir_format
from backend.api.db.models.notebook import Notebook
from backend.api.db.models.validator import (
    date_validator,
    parse_bool_validator,
    phone_validator,
    postal_code_validator,
)


class BeneficiaryStructure(BaseModel):
    structure_id: UUID
    structure_name: str
    beneficiary_status: str


class Beneficiary(BaseModel):
    id: UUID
    email: str | None
    lastname: str
    firstname: str
    caf_number: str | None
    pe_number: str | None
    postal_code: str | None
    city: str | None
    address1: str | None
    address2: str | None
    mobile_number: str | None
    date_of_birth: date
    place_of_birth: str | None
    deployment_id: UUID
    created_at: datetime
    updated_at: datetime
    internal_id: str | None
    notebook: Notebook | None
    # BRSA users may not have an account
    # (account is created on the first login attempt)
    account_id: UUID | None
    nir: str | None
    pe_unique_import_id: str | None


def snake_to_camel(field):
    parts = field.split("_")
    return "".join(parts[0:1] + [part.title() for part in parts[1:]])


def is_same_name(firstname1, firstname2, lastname1, lastname2):
    return (
        firstname1.lower().strip() == firstname2.lower().strip()
        and lastname1.lower().strip() == lastname2.lower().strip()
    )


class BeneficiaryImport(BaseModel):
    internal_id: str = Field(..., alias="Identifiant dans le SI*")
    firstname: str = Field(..., alias="Prénom*")
    lastname: str = Field(..., alias="Nom*")
    date_of_birth: date = Field(..., alias="Date de naissance*")
    place_of_birth: str | None = Field(None, alias="Lieu de naissance")
    mobile_number: str | None = Field(None, alias="Téléphone")
    email: EmailStr | None = Field(None, alias="Email")
    address1: str | None = Field(None, alias="Adresse")
    address2: str | None = Field(None, alias="Adresse (complément)")
    postal_code: str | None = Field(None, alias="Code postal")
    city: str | None = Field(None, alias="Ville")
    work_situation: str | None = Field(None, alias="Situation")
    caf_number: str | None = Field(None, alias="Numéro allocataire CAF/MSA")
    pe_number: str | None = Field(None, alias="Identifiant Pôle emploi")
    right_rsa: str | None = Field(None, alias="Droits RSA")
    right_are: bool | None = Field(None, alias="Droits ARE")
    right_ass: bool | None = Field(None, alias="Droits ASS")
    right_bonus: bool | None = Field(None, alias="Prime d'activité")
    right_rqth: bool | None = Field(None, alias="RQTH")
    geographical_area: str | None = Field(None, alias="Zone de mobilité")
    rome_code_description: str | None = Field(
        None, alias="Emploi recherché (code ROME)"
    )
    education_level: str | None = Field(None, alias="Niveau de formation")
    structure_name: str | None = Field(None, alias="Structure")
    advisor_email: EmailStr | None = Field(None, alias="Accompagnateurs")
    nir: str | None = Field(None, alias="NIR")

    class Config:
        anystr_strip_whitespace = True
        allow_population_by_field_name = True

    _phone_validator = phone_validator("mobile_number")
    _postal_code_validator = postal_code_validator("postal_code")
    _parse_bool_validator = parse_bool_validator(
        "right_are", "right_ass", "right_bonus", "right_rqth", pre=True
    )
    _date_validator = date_validator("date_of_birth", pre=True)

    @validator("right_rsa")
    def parse_right_rsa(cls, right_rsa):
        if right_rsa and right_rsa not in [
            "rsa_droit_ouvert_et_suspendu",
            "rsa_droit_ouvert_versable",
            "rsa_droit_ouvert_versement_suspendu",
        ]:
            raise ValueError("value unknown")
        return right_rsa

    @validator("geographical_area")
    def parse_geographical_area(cls, geographical_area):
        if geographical_area and geographical_area not in [
            "none",
            "less_10",
            "between_10_20",
            "between_20_30",
            "plus_30",
        ]:
            raise ValueError("value unknown")
        return geographical_area

    @validator("work_situation")
    def parse_work_situation(cls, work_situation):
        if work_situation and work_situation not in [
            "recherche_emploi",
            "recherche_formation",
            "recherche_alternance",
            "recherche_stage",
            "emploi",
            "formation",
            "stage",
            "alternance",
            "service_civique",
            "iae",
            "interim",
            "construction_projet",
            "projet_entrepreneurial",
            "entrepreneur",
            "etudiant",
            "scolarisé",
            "enretraite",
            "maladie",
            "invalidite",
            "conge_parental",
            "au_foyer",
            "autre",
            "cdi_temps_plein",
            "cdi_temps_partiel",
            "cdd_temps_plein",
            "cdd_temps_partiel",
            "intermittent",
        ]:
            raise ValueError("value unknown")
        return work_situation

    @validator("education_level")
    def parse_education_level(cls, education_level):
        if education_level and education_level not in [
            "AFS",
            "C12",
            "C3A",
            "CFG",
            "CP4",
            "NV5",
            "NV4",
            "NV3",
            "NV2",
            "NV1",
        ]:
            raise ValueError("value unknown")
        return education_level

    @validator("nir")
    def parse_nir(cls, nir: str):
        validation_error = nir_format(nir)
        if validation_error:
            raise ValueError(validation_error)
        return nir

    # TODO Should be possible to use decorator to mark field for beneficiary
    def get_beneficiary_editable_keys(self) -> list[str]:
        BENEFICIARY_EDITABLE_FIELDS = [
            "place_of_birth",
            "mobile_number",
            "address1",
            "address2",
            "postal_code",
            "city",
            "caf_number",
            "pe_number",
            "email",
            "nir",
        ]
        return self._filter_updatable_field_keys(BENEFICIARY_EDITABLE_FIELDS)

    # TODO Should be possible to use decorator to mark field for notebook
    def get_notebook_editable_keys(self) -> list[str]:
        NOTEBOOK_EDITABLE_FIELDS = [
            "right_rsa",
            "right_rqth",
            "right_are",
            "right_ass",
            "right_bonus",
            "work_situation",
            "education_level",
            "geographical_area",
        ]
        return self._filter_updatable_field_keys(NOTEBOOK_EDITABLE_FIELDS)

    def _filter_updatable_field_keys(self, allowed_fields) -> list[str]:
        filled_fields = self._get_filled_field_keys()
        keys_to_update = [key for key in filled_fields if key in allowed_fields]
        # Special case for address where we need to clean address2 if address1 is provided but not address2
        if "address1" in keys_to_update and "address2" not in keys_to_update:
            keys_to_update.append("address2")

        return keys_to_update

    def get_values_for_keys(self, keys: list[str]) -> list[object]:
        return [self.dict().get(key) for key in keys]

    def _get_filled_field_keys(self):
        return [
            fieldname for (fieldname, value) in self.dict().items() if value is not None
        ]


class BeneficiaryCsvRowResponse(BaseModel):
    row: dict | None = None
    data: BeneficiaryImport | None = None
    valid: bool
    update: bool = False
    errors: list[CsvFieldError] | None = None


def map_csv_row(row: Series) -> BeneficiaryCsvRowResponse:
    try:
        data = BeneficiaryImport.parse_obj(row)
        return BeneficiaryCsvRowResponse(data=data, valid=True)

    except ValidationError as error:
        return BeneficiaryCsvRowResponse(
            row=row.to_dict(),
            valid=False,
            errors=[
                CsvFieldError(key="".join(str(err["loc"][0])), error=err["msg"])
                for err in error.errors()
            ],
        )
