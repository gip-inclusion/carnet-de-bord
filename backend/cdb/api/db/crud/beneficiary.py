import logging
from datetime import date
from typing import List
from uuid import UUID

from asyncpg import Record
from asyncpg.connection import Connection
from gql.dsl import DSLField, DSLSchema

from cdb.api.core.exceptions import InsertFailError
from cdb.api.db.crud.notebook import (
    NOTEBOOK_BASE_FIELDS,
    NOTEBOOK_BASE_JOINS,
    insert_notebook,
    insert_notebook_member,
    parse_notebook_from_record,
)
from cdb.api.db.crud.notebook_info import insert_or_update_need_orientation
from cdb.api.db.crud.professional import get_professional_by_email
from cdb.api.db.crud.professional_project import insert_professional_projects
from cdb.api.db.crud.structure import get_structure_by_name
from cdb.api.db.models.beneficiary import (
    Beneficiary,
    BeneficiaryImport,
    BeneficiaryStructure,
    BeneficiaryWithAdminStructureEmail,
)
from cdb.api.db.models.deployment import parse_deployment_from_beneficiary_record
from cdb.api.db.models.notebook_member import NotebookMemberInsert
from cdb.api.db.models.orientation_system import OrientationSystem
from cdb.api.db.models.professional import Professional
from cdb.api.v1.payloads.notebook import NotebookInput

logger = logging.getLogger(__name__)

BENEFICIARY_BASE_QUERY = (
    """
    SELECT b.*, acc.id as account_id,
    dep.id as dep_id, dep.label as dep_label, dep.department_code as dep_code,
    """
    + NOTEBOOK_BASE_FIELDS
    + """
    FROM public.beneficiary b
    LEFT JOIN notebook n ON n.beneficiary_id = b.id
    LEFT JOIN deployment dep on dep.id = b.deployment_id
    """
    + NOTEBOOK_BASE_JOINS
    + """
    LEFT JOIN public.account acc
    ON acc.beneficiary_id = b.id
"""
)


# see https://betakuang.medium.com/why-postgresqls-on-conflict-cannot-find-my-partial-unique-index-552327b85e1
firstname_lastname_date_of_birth_unique_idx = (
    "(LOWER(trim(firstname)), LOWER(trim(lastname)), date_of_birth, deployment_id)"
)


async def get_beneficiaries_like(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
) -> list[Beneficiary]:
    matching_beneficiaries: list[Record] = await connection.fetch(
        """
        SELECT *
        FROM beneficiary
        WHERE (
            lower(trim(firstname)) = lower(trim($1))
            AND lower(trim(lastname)) = lower(trim($2))
            AND date_of_birth = $3 AND deployment_id = $5
        ) OR (internal_id = $4 AND deployment_id = $5)
        """,
        beneficiary.firstname,
        beneficiary.lastname,
        beneficiary.date_of_birth,
        beneficiary.internal_id,
        deployment_id,
    )
    return [
        Beneficiary.parse_obj(beneficiary) for beneficiary in matching_beneficiaries
    ]


async def update_beneficiary_field(
    connection: Connection,
    field_name: str,
    field_value: str,
    beneficiary_id: UUID,
) -> UUID | None:
    result: Record | None = await connection.fetchrow(
        f"""
UPDATE beneficiary SET {field_name} = $2
where id = $1
returning id
        """,
        beneficiary_id,
        field_value,
    )
    if result:
        return result["id"]


async def update_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    beneficiary_id: UUID,
) -> UUID | None:
    keys_to_update = beneficiary.get_beneficiary_editable_keys()

    if len(keys_to_update) == 0:
        logger.info("trying to update beneficiary but no fields where updated.")
        return beneficiary_id

    sql_fields = [f"{key} = ${index+2}" for (index, key) in enumerate(keys_to_update)]

    sql = f"""
        UPDATE beneficiary
        SET {", ".join(sql_fields)}
        WHERE id=$1
            returning id
"""

    result = await connection.fetchrow(
        sql,
        beneficiary_id,
        *beneficiary.get_values_for_keys(keys_to_update),
    )

    if result:
        return result["id"]


async def insert_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
) -> UUID | None:
    mandatory_keys = ["firstname", "lastname", "internal_id", "date_of_birth"]
    keys_to_insert = mandatory_keys + beneficiary.get_beneficiary_editable_keys()
    sql_values = beneficiary.get_values_for_keys(keys_to_insert)
    keys_to_insert.append("deployment_id")
    sql_values.append(deployment_id)
    sql = f"""
        INSERT INTO public.beneficiary ({", ".join(keys_to_insert)})
        VALUES ({", ".join([f"${x+1}" for x in range(len(keys_to_insert))])})
            returning id
"""
    created_beneficiary: Record = await connection.fetchrow(sql, *sql_values)

    if created_beneficiary:
        return created_beneficiary["id"]


async def get_beneficiary_with_query(
    connection: Connection, query: str, *args
) -> Beneficiary | None:
    async with connection.transaction():
        beneficiary = None

        beneficiary_records: List[Record] = await connection.fetch(
            BENEFICIARY_BASE_QUERY + query,
            *args,
        )

        for beneficiary_record in beneficiary_records:
            if beneficiary is None:
                beneficiary = Beneficiary.parse_obj(beneficiary_record)
                beneficiary.deployment = parse_deployment_from_beneficiary_record(
                    beneficiary_record
                )
            if beneficiary_record["n_id"] is not None:
                beneficiary.notebook = await parse_notebook_from_record(
                    beneficiary_record,
                    notebook=beneficiary.notebook,
                )

        return beneficiary


async def get_beneficiary_from_personal_information(
    connection: Connection, firstname: str, lastname: str, birth_date: date
) -> Beneficiary | None:
    return await get_beneficiary_with_query(
        connection,
        # We use LOWER(BTRIM(field)) here to match the index so postgres
        # can use the index rather than scanning the whole table
        "WHERE LOWER(BTRIM(b.firstname)) = LOWER($1) AND "
        "LOWER(BTRIM(b.lastname)) = LOWER($2) AND "
        "b.date_of_birth = $3",
        firstname,
        lastname,
        birth_date,
    )


async def add_beneficiary_to_structure(
    connection: Connection,
    beneficiary_id: UUID,
    structure_id: UUID,
    status: str,
) -> UUID | None:
    result = await connection.fetchrow(
        """
INSERT INTO beneficiary_structure (beneficiary_id, structure_id, status)
VALUES ($1, $2, $3)
RETURNING id
        """,
        beneficiary_id,
        structure_id,
        status,
    )
    if result:
        return result["id"]


async def get_structures_for_beneficiary(
    connection: Connection,
    beneficiary_id: UUID,
) -> list[BeneficiaryStructure]:
    rows: list[Record] = await connection.fetch(
        """
SELECT struct.id, struct.name, b_struct.status
FROM public.beneficiary_structure AS b_struct
LEFT JOIN public.structure AS struct
    ON struct.id = b_struct.structure_id
WHERE b_struct.beneficiary_id = $1
        """,
        beneficiary_id,
    )
    return [
        BeneficiaryStructure(
            structure_id=row["id"],
            structure_name=row["name"],
            beneficiary_status=row["status"],
        )
        for row in rows
    ]


async def get_beneficiary_by_id(
    connection: Connection, beneficiary_id: UUID
) -> Beneficiary | None:
    return await get_beneficiary_with_query(
        connection, "WHERE b.id = $1", beneficiary_id
    )


async def create_beneficiary_with_notebook_and_referent(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id: UUID,
    need_orientation: bool,
) -> UUID:
    beneficiary_id: UUID | None = await insert_beneficiary(
        connection, beneficiary, deployment_id
    )

    if not beneficiary_id:
        raise InsertFailError("insert beneficiary failed")

    new_notebook_id: UUID | None = await insert_notebook(
        connection, beneficiary_id, beneficiary
    )

    if not new_notebook_id:
        raise InsertFailError("insert notebook failed")

    await insert_or_update_need_orientation(
        connection, new_notebook_id, None, need_orientation
    )

    await insert_professional_projects(connection, new_notebook_id, beneficiary)
    await add_referent_and_structure_to_beneficiary(
        connection,
        beneficiary_id,
        new_notebook_id,
        beneficiary,
        deployment_id,
    )
    logger.info("inserted new beneficiary %s", beneficiary_id)
    return beneficiary_id


async def add_referent_and_structure_to_beneficiary(
    connection: Connection,
    beneficiary_id: UUID,
    notebook_id: UUID,
    beneficiary: BeneficiaryImport,
    deployment_id: UUID,
):
    structure = None
    if beneficiary.structure_name:
        structure = await get_structure_by_name(
            connection, beneficiary.structure_name, deployment_id
        )
        if structure is None:
            logger.info(
                "Trying to associate structure with beneficiary: "
                'structure "%s" does not exist',
                beneficiary.structure_name,
            )
            return
    referent = None
    if beneficiary.advisor_email:
        referent: Professional | None = await get_professional_by_email(
            connection, beneficiary.advisor_email.strip()
        )
    if structure:
        await add_beneficiary_to_structure(
            connection, beneficiary_id, structure.id, "current"
        )
    elif referent:
        await add_beneficiary_to_structure(
            connection, beneficiary_id, referent.structure_id, "current"
        )

    if referent and referent.account_id:
        # Si on a pu récupérer le compte du référent fourni, on l'ajoute dans le
        # groupe de suivi en tant que référent.
        if not structure or structure.id == referent.structure_id:
            referent_member = NotebookMemberInsert(
                notebook_id=notebook_id,
                account_id=referent.account_id,
                member_type="referent",
            )
            await insert_notebook_member(connection, referent_member)
    else:
        # Si un référent est fourni mais qu'on ne le connaît pas, on ne fait rien.
        logger.info(
            "trying to create referent: no account with email: %s",
            beneficiary.advisor_email,
        )


async def get_beneficiaries_without_referent(
    connection: Connection,
) -> list[BeneficiaryWithAdminStructureEmail]:
    rows: list[Record] = await connection.fetch(
        """
SELECT b.firstname, b.lastname, b.date_of_birth,
    os.id as orientation_id, os.name as orientation_name,
    os.orientation_type as orientation_type,
    os.created_at as orientation_created_at,
    os.updated_at as orientation_updated_at,
    ads.email AS admin_structure_email, s.name AS structure_name
FROM beneficiary AS b
LEFT JOIN notebook AS n ON n.beneficiary_id = b.id
LEFT JOIN notebook_info AS ni ON ni.notebook_id = n.id
LEFT JOIN (
    SELECT * FROM notebook_member AS nm
    WHERE nm.active AND nm.member_type = 'referent'
    ) referent ON referent.notebook_id = n.id
LEFT JOIN orientation_system AS os ON os.id = ni.orientation_system_id
LEFT JOIN account_info  AS ai ON referent.account_id = ai.account_id
LEFT JOIN beneficiary_structure bs ON bs.beneficiary_id = b.id
LEFT JOIN admin_structure_structure ass ON ass.structure_id = bs.structure_id
LEFT JOIN admin_structure ads ON ads.id = ass.admin_structure_id
LEFT JOIN structure s ON s.id = ass.structure_id
WHERE referent IS NULL AND ads.email IS NOT NULL AND status = 'current'
ORDER BY ads.email ASC, s.name ASC, firstname ASC, lastname ASC, date_of_birth ASC;
        """
    )
    return [
        BeneficiaryWithAdminStructureEmail(
            admin_structure_email=row["admin_structure_email"],
            structure_name=row["structure_name"],
            firstname=row["firstname"],
            lastname=row["lastname"],
            date_of_birth=row["date_of_birth"],
            orientation=OrientationSystem(
                id=row["orientation_id"],
                name=row["orientation_name"],
                orientation_type=row["orientation_type"],
                created_at=row["orientation_created_at"],
                updated_at=row["orientation_updated_at"],
            )
            if row["orientation_id"] is not None
            else None,
        )
        for row in rows
    ]


def get_insert_beneficiary_mutation(
    dsl_schema: DSLSchema,
    deployment_id: str,
    notebook: NotebookInput,
) -> dict[str, DSLField]:
    """
    TODO:
        Création du bénéficiaire et du notebook avec la mutation suivante :
        mutation CreateBeneficiaryWithNotebook(
          $nir: String!
          $deploymentId: uuid!
          $firstname: String!
          $lastname: String!
          $dateOfBirth: date!
            $internalId: String
          $mobileNumber: String
            $email: citext
            $address1: String
          $address2: String
          $postalCode: String
            $city: String
            $cafNumber: String
        ) {
            insert_beneficiary_one(object: {
              nir: $nir
              deploymentId: $deploymentId
              firstname: $firstname
              lastname: $lastname
                dateOfBirth: $dateOfBirth
                    internalId: $internalId
                mobileNumber: $mobileNumber
                    email: $email
                    address1: $address1
                address2: $address2
                postalCode: $postalCode
                    city: $city
                    cafNumber: $cafNumber
              notebook: { data: {} }
            }) {
                notebook {
                    id
                }
          }
        }

        Question : quid de la structure ?
         (comment la trouver et créer la beneficiary_structure associée)
    """

    return {
        "create_beneficiary_with_notebook": (
            dsl_schema.mutation_root.insert_beneficiary_one.args(
                object={
                    "nir": notebook.nir,
                    "firstname": notebook.firstname,
                    "lastname": notebook.lastname,
                    "dateOfBirth": notebook.date_of_birth,
                    #    internalId: $internalId
                    # mobileNumber: $mobileNumber
                    #    email: $email
                    #    address1: $address1
                    # address2: $address2
                    # postalCode: $postalCode
                    #    city: $city
                    #    cafNumber: $cafNumber
                    "notebook": {"data": {}},
                }
            ).select(dsl_schema.beneficiary.notebook.select(dsl_schema.notebook.id))
        ),
    }
