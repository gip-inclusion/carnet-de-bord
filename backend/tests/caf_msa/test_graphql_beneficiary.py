from datetime import date
from uuid import UUID

from asyncpg.connection import Connection
from gql.client import AsyncClientSession

from cdb.api.db.crud.external_data import get_external_data_with_query
from cdb.api.db.graphql.beneficiary import (
    get_beneficiary_by_nir,
    update_beneficiary,
)
from cdb.api.db.models.external_data import ExternalData
from cdb.caf_msa.parse_infos_foyer_rsa import CdbBeneficiaryInfos


async def test_get_beneficiary_by_nir(
    gql_manager_client: AsyncClientSession,
    sophie_tifour_beneficiary_id: UUID,
    sophie_tifour_nir: str,
):
    beneficiary = await get_beneficiary_by_nir(gql_manager_client, sophie_tifour_nir)

    assert beneficiary
    assert beneficiary.id == sophie_tifour_beneficiary_id
    assert beneficiary.right_rsa == "rsa_droit_ouvert_et_suspendu"
    assert beneficiary.rsa_closure_reason is None
    assert beneficiary.rsa_closure_date is None
    assert beneficiary.rsa_suspension_reason is None
    assert beneficiary.is_homeless is None
    assert beneficiary.subject_to_right_and_duty is None
    assert beneficiary.caf_number == "2055990"


async def test_update_beneficiary_by_id(
    gql_manager_client: AsyncClientSession, sophie_tifour_beneficiary_id: UUID
):
    personne = CdbBeneficiaryInfos(
        right_rsa="rsa_droit_ouvert_et_suspendu",
        rsa_closure_date=None,
        rsa_closure_reason=None,
        rsa_suspension_reason="caf_ressources_trop_eleve",
        is_homeless=False,
        subject_right_and_duty=False,
        caf_number="2055990",
    )

    sha = "123"
    external_data_payload = {"payload": "payload"}
    beneficiary = await update_beneficiary(
        gql_manager_client,
        sophie_tifour_beneficiary_id,
        personne,
        sha,
        external_data_payload,
    )
    assert beneficiary
    assert beneficiary.right_rsa == "rsa_droit_ouvert_et_suspendu"
    assert beneficiary.rsa_closure_reason is None
    assert beneficiary.rsa_closure_date is None
    assert beneficiary.rsa_suspension_reason == "caf_ressources_trop_eleve"
    assert beneficiary.is_homeless is False
    assert beneficiary.subject_to_right_and_duty is False


async def test_update_beneficiary_create_external_data(
    gql_manager_client: AsyncClientSession,
    db_connection: Connection,
    sophie_tifour_beneficiary_id: UUID,
    snapshot,
):
    personne = CdbBeneficiaryInfos(
        right_rsa="rsa_droit_ouvert_et_suspendu",
        rsa_closure_date=None,
        rsa_closure_reason=None,
        rsa_suspension_reason="caf_ressources_trop_eleve",
        is_homeless=False,
        subject_right_and_duty=False,
        caf_number="2055990",
    )

    sha = "123"
    external_data_payload = {"payload": "payload", "madate": date(2003, 2, 1)}
    beneficiary = await update_beneficiary(
        gql_manager_client,
        sophie_tifour_beneficiary_id,
        personne,
        sha,
        external_data_payload,
    )
    assert beneficiary
    external_data: ExternalData | None = await get_external_data_with_query(
        db_connection,
        "WHERE external_data_info.beneficiary_id = $1 "
        "AND external_data.source = $2 "
        "ORDER BY created_at DESC LIMIT 1",
        beneficiary.id,
        "cafmsa",
    )
    assert external_data
    assert external_data.data == snapshot


async def test_update_beneficiary_rsa_closure(
    gql_manager_client: AsyncClientSession, sophie_tifour_beneficiary_id: UUID
):
    personne = CdbBeneficiaryInfos(
        right_rsa="rsa_clot",
        rsa_closure_date="2023-01-01",
        rsa_closure_reason="caf_regroupement",
        rsa_suspension_reason=None,
        is_homeless=True,
        subject_right_and_duty=True,
        caf_number="XXXXXXXX",
    )
    sha = "123"
    external_data = {"payload": "payload"}
    beneficiary = await update_beneficiary(
        gql_manager_client,
        sophie_tifour_beneficiary_id,
        personne,
        sha,
        external_data,
    )
    assert beneficiary
    assert beneficiary.right_rsa == "rsa_clot"
    assert beneficiary.rsa_closure_reason == "caf_regroupement"
    assert beneficiary.rsa_closure_date == date(2023, 1, 1)
    assert beneficiary.rsa_suspension_reason is None
    assert beneficiary.is_homeless is True
    assert beneficiary.subject_to_right_and_duty is True
    assert beneficiary.caf_number == "XXXXXXXX"


async def test_update_beneficiary_rsa_suspension(
    gql_manager_client: AsyncClientSession,
    sophie_tifour_beneficiary_id: UUID,
):
    personne = CdbBeneficiaryInfos(
        right_rsa="rsa_droit_ouvert_et_suspendu",
        rsa_closure_date=None,
        rsa_closure_reason=None,
        rsa_suspension_reason="caf_ressources_trop_eleve",
        is_homeless=False,
        subject_right_and_duty=True,
        caf_number="XXXXXXXX",
    )
    sha = "123"
    external_data = {"payload": "payload", "infos": {"madate": date(2003, 2, 1)}}
    beneficiary = await update_beneficiary(
        gql_manager_client,
        sophie_tifour_beneficiary_id,
        personne,
        sha,
        external_data,
    )
    assert beneficiary
    assert beneficiary.right_rsa == "rsa_droit_ouvert_et_suspendu"
    assert beneficiary.rsa_closure_reason is None
    assert beneficiary.rsa_closure_date is None
    assert beneficiary.rsa_suspension_reason == "caf_ressources_trop_eleve"
    assert beneficiary.is_homeless is False
    assert beneficiary.subject_to_right_and_duty is True
    assert beneficiary.caf_number == "XXXXXXXX"
