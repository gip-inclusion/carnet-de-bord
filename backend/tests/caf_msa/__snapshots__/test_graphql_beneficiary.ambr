# serializer version: 1
# name: test_update_beneficiary_create_external_data
  dict({
    'madate': '2003-02-01',
    'payload': 'payload',
  })
# ---
